teal/packer/async.tl

```diff
@@ -0,0 +1,143 @@
+local record Async
+  type Thunk = function(function(...:any))
+
+  -- Order by highest number of return types
+
+  void: function             (function()           ): function()
+  void: function<A1>         (function(A1)         ): function(A1)
+  void: function<A1,A2>      (function(A1,A2)      ): function(A1,A2)
+  void: function<A1,A2,A3,A4>(function(A1,A2,A3,A4)): function(A1,A2,A3,A4)
+
+  wrap: function<A1,A2,R1,R2>      (function(A1,A2,      function(R1,R2)), integer): function(A1,A2         ): R1,R2
+  wrap: function<A1,A2,R1>         (function(A1,A2,      function(R1)   ), integer): function(A1,A2         ): R1
+  wrap: function<A1,A2,A3,R1,R2>   (function(A1,A2,A3,   function(R1,R2)), integer): function(A1,A2,A3      ): R1,R2
+  wrap: function<A1,A2,A3,A4,R1>   (function(A1,A2,A3,A4,function(R1)   ), integer): function(A1,A2,A3,A4   ): R1
+  wrap: function<A1,A2,A3,R1>      (function(A1,A2,A3,   function(R1)   ), integer): function(A1,A2,A3      ): R1
+  wrap: function<A1,A2,A3,A4,A5,R1>(function(A1,A2,A3,A4,A5,function(R1)), integer): function(A1,A2,A3,A4,A5): R1
+  wrap: function<A1,A2,A3>         (function(A1,A2,A3,   function())     , integer): function(A1,A2,A3      )
+
+  sync: function<T>(T, integer): T
+
+  main: function()
+  join: function(integer, (function(): boolean), {Thunk}): {any}
+  curry: function(fn: function, ...: any): Async.Thunk
+end
+
+local co = coroutine
+
+---Executes a future with a callback when it is done
+local function execute(func: function, callback: function, ...: any)
+  local thread = co.create(func)
+
+  local function step(...: any)
+    local ret = {co.resume(thread, ...)}
+    local stat, nargs, fn_or_ret = unpack(ret) as (boolean, integer, function)
+
+    if not stat then
+      error(string.format("The coroutine failed with this message: %s\n%s",
+        nargs, debug.traceback(thread)))
+    end
+
+    if co.status(thread) == 'dead' then
+      if callback then
+        callback(unpack(ret, 3))
+      end
+      return
+    end
+
+    local args = {select(4, unpack(ret))}
+    args[nargs] = step
+    fn_or_ret(unpack(args, 1, nargs))
+  end
+
+  step(...)
+end
+
+local record M
+  main: function
+end
+
+---Creates an async function with a callback style function.
+---
+---@generic F : function
+function M.wrap(func: function, argc: integer): function
```

Comment:
I spent a lot of time trying to avoid this.

Using `nparams` doesn't work for wrapped functions. (E.g. `vim.schedule_wrap` or the debouncing functions used in `Gitsigns`), since they use `...` to pass through the arguments.

Additionally, the Lua reference manual explicitly recommends not to use `debug` module in none debug applications. Though the case might be different for LuaJIT.

I think there are other issues, I can't remember.

I'm pretty confident that this interface is the best we are going to be able to achieve.

Reply:
Fair enough. I would still love to be able to get rid of the extra param, but I trust you've thought through the alternatives thoroughly.